src/client/clientmap.cpp

```diff
@@ -180,17 +204,420 @@ void ClientMap::getBlocksInViewRange(v3s16 cam_pos_nodes,
+				continue; // Out of range, skip.
+
+			// Frustum culling
+			// Only do coarse culling here, to account for fast camera movement.
+			// This is needed because this function is not called every frame.
+			float frustum_cull_extra_radius = 300.0f;
+			if (is_frustum_culled(mesh_sphere_center,
+					mesh_sphere_radius + frustum_cull_extra_radius))
+				continue;
+
+			// Calculate the vector from the camera block to the current block
+			// We use it to determine through which sides of the current block we can continue the search
+			v3s16 look = block_coord - camera_block;
+
+			// Occluded near sides will further occlude the far sides
+			u8 visible_outer_sides = flags & 0x07;
+
+			// Raytraced occlusion culling - send rays from the camera to the block's corners
+			if (occlusion_culling_enabled && m_enable_raytraced_culling &&
+					block && mesh &&
+					visible_outer_sides != 0x07 && isBlockOccluded(block, cam_pos_nodes)) {
+				blocks_occlusion_culled++;
+				continue;
+			}
+
+			// The block is visible, add to the draw list
+			if (mesh) {
+				// Add to set
+				block->refGrab();
+				m_drawlist[block_coord] = block;
+			}
+
+			// Decide which sides to traverse next or to block away
+
+			// First, find the near sides that would occlude the far sides
+			// * A near side can itself by occluded a nearby block (the test above ^^)
+			// * A near side can be visible but fully opaque by itself (e.g. ground at the 0 level)
+
+			// mesh solid sides are +Z-Z+Y-Y+X-X
+			// do not test sides where we are 'inside' the block's coordinates
+			u8 ignore_inner_sides = (look.X == 0 ? 3 : 0) |
+				(look.Y == 0 ? 12 : 0) |
+				(look.Z == 0 ? 48 : 0);
+
+			// get the mask for the sides that are relevant based on the direction
+			u8 near_inner_sides = (look.X > 0 ? 1 : 2) |
+					(look.Y > 0 ? 4 : 8) |
+					(look.Z > 0 ? 16 : 32);
+			
+			// This bitset is +Z-Z+Y-Y+X-X (See MapBlockMesh), and axis is XYZ.
+			u8 transparent_sides = (occlusion_culling_enabled && block) ? ~block->solid_sides : 0x3F;
+
+			// compress block transparent sides to ZYX mask of see-through axes
+			u8 near_transparency = ((transparent_sides & near_inner_sides) | ignore_inner_sides) & 0x3F;
+
+			near_transparency = (near_transparency | (near_transparency >> 1)) & 0x15;
+			near_transparency = (near_transparency & 1) |
+					((near_transparency >> 1) & 2) |
+					((near_transparency >> 2) & 4);
+
+			// combine with known visible sides that matter
+			near_transparency &= visible_outer_sides;
+
+			// The rule for any far side to be visible:
+			// * Any of the adjacent near sides is transparent (different axes)
+			// * The opposite near side (same axis) transparent, and is the dominant axis of the look vector
+
+			// dominant axis flag
+			u8 dominant_axis = (abs(look.X) > abs(look.Y) && abs(look.X) > abs(look.Z)) |
+						((abs(look.Y) > abs(look.Z) && abs(look.Y) > abs(look.Z)) << 1) |
+						((abs(look.Z) > abs(look.X) && abs(look.Z) > abs(look.Y)) << 2);
```

Comment:
2nd line has typo.

Just keep this simple.
```suggestion
			u8 dominant_axis = abs(look.X) > abs(look.Y) ?
				(abs(look.X) > abs(look.Z) ? 1 : 4) :
				(abs(look.Y) > abs(look.Z) ? 2 : 4);
```

Reply:
Good catch with the typo.

The dominant axis is only dominant if it's the only one with maximum value. If all three axes are equal, or if two are dominant candidates then there's no dominant axis and the far side is occluded.

Reply:
There's always a dominant axis if floating-point camera coordinates are used.

And if you really want to keep `look` as blockpos dir, you'd basically have multiple dominant axes, because the camera can be anywhere inside the block.